core/Segment.php

```diff
@@ -460,8 +465,42 @@ public function getHash()
 
     public static function getSegmentHash($definition)
     {
-        // urldecode to normalize the string, as browsers may send slightly different payloads for the same archive
-        return md5(urldecode($definition));
+        $cache = Cache::getEagerCache();
+        $cacheKey = self::CACHE_KEY . md5($definition);
+
+        if ($cache->contains($cacheKey)) {
+            return $cache->fetch($cacheKey);
+        }
+
+        $defaultHash = md5(urldecode($definition));
+
+        // if the cache for segments already built, but this segment was not found,
+        // we return the default segment, this can be a segment from url or
+        // something like "visitorType==new"
+        if ($cache->contains(self::SEGMENT_HAS_BUILT_CACHE_KEY)) {
+            return $defaultHash;
+        }
+
+        // the segment hash is not built yet, let's do it
+        $cache->save(self::SEGMENT_HAS_BUILT_CACHE_KEY, true);
```

Comment:
just to be safe we would maybe want to set this one after the `for` loop and after all cache entries were set correctly. Just in case there was some exception in the for loop and we would maybe then write a partial cache only but assume we have written the full cache. not sure what would happen in case of an exception etc and if it's even possible but just meaning to be safe.

Reply:
make sense 👍